pennylane/ops/op_math/symbolicop.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule defines a base class for symbolic operations representing operator math.
+"""
+from copy import copy
+
+from pennylane.operation import Operator
+from pennylane.queuing import QueuingContext
+
+
+class SymbolicOp(Operator):
+    """Developer-facing base class for single-operator symbolic operators.
+
+    Args:
+        base (~.operation.Operator): The base operation that is modified symbolicly
+        do_queue (bool): indicates whether the operator should be
+            recorded when created in a tape context
+        id (str): custom label given to an operator instance,
+            can be useful for some applications where the instance has to be identified
+
+    This *developer-facing* class can serve as a parent to single base symbolic operators, such as
+    :class:`~.ops.op_math.Adjoint` and :class:`~.ops.op_math.Pow`.
+
+    New symbolic operators can inherit from this class to recieve some common default behavior, such as
+    deferring properties to the the base class, copying the base class during a shallow copy, and updating
+    the metadata of the base operator during queueing.
+
+    The child symbolic operator should define the `_name` property during initialization and define any
+    relevant representations, such as :meth:`~.operation.Operator.matrix`, :meth:`~.operation.Operator.diagonalizing_gates`,
+    :meth:`~.operation.Operator.eigvals`, and :meth:`~.operation.Operator.decomposition`.
+    """
+
+    _name = "Symbolic"
```

Comment:
Why define the name like this? Is it to ensure that every instance of `SymbolicOp` will have the same `cls._name` attribute?

Reply:
This needed to be defined so that we could instantiate the class and inspect it without an error occurring.  Child classes should override this in their `__init__` function. I didn't want to define this during `SymbolicOp.__init__`, since I didn't want to accidentally override the `_name` property of the child class.